sklearn/ensemble/_bagging.py

```diff
@@ -46,6 +46,65 @@
 MAX_INT = np.iinfo(np.int32).max
 
 
+def _get_n_samples_bootstrap(n_samples, max_samples, sample_weight):
+    """
+    Get the number of samples in a bootstrap sample.
+
+    Parameters
+    ----------
+    n_samples : int
+        Number of samples in the dataset.
+
+    max_samples : None, int or float
+        The maximum number of samples to draw.
+
+        - If None, then draw `n_samples` samples.
+        - If int, then draw `max_samples` samples.
+        - If float, then draw `max_samples * n_samples` unweighted samples or
+          `max_samples * sample_weight.sum()` weighted samples.
+
+    sample_weight : array of shape (n_samples,) or None
+        Sample weights with frequency semantics when `max_samples` is explicitly
+        set to a float or integer value. When keeping the `max_samples=None` default
+        value, the equivalence between fitting with integer weighted data points or
+        integer repeated data points is no longer guaranteed because the effective
+        bootstrap size is no longer guaranteed to be equivalent.
```

Comment:
For reviewers, I updated the notebooks of our sample weight analyzer, in particular this want to highlight the fact that `max_samples=None` breaks frequency weights semantics:

https://github.com/snath-xoc/sample-weight-audit-nondet/blob/main/reports/sklearn_estimators_sample_weight_audit_investigative_plots.ipynb

The p-value is very small: the null-hypothesis that fitting with integer weights is equivalent to fitting with repeated data points is rejected.

Editing the cell to select either `max_samples=1.0` or `max_samples=100` restores large p-values, even when increasing `n_stochastic_fits` meaning that we cannot reject the null-hypothesis that frequency weights semantics are respected.